### PR TITLE
[core] Text documents

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Draft, leaving this as a reminder

(this now contains #3892)

## Changelog

* pmd-apex
  * Text position handling is simplified (SourceCodePositioner is now part of TextDocument)
  * The Location attribute is removed
  * EmptyCatchBlock now considers doc comments as regular comments
  * Some nodes now have a different report location: eg for methods and classes it's just their identifier, like for pmd-java, ref #730 
  
* pmd-core
  * New API: Chars, a string slice type. Supports efficient string-like operations without copying the char array of the file. Makes comment handling more efficient, in a future PR, will be used by Javacc tokens.
  * TextFile is improved
    * Line terminators are normalized to `\n`
    * Reading the file computes a checksum which is used by incremental analysis later 
    * TextFile replaces DataSource everywhere in the internals of PMD
  * New API: TextDocument, built on top of TextFile
  * New API: FileLocation. This represents the line/column coordinates of a place in a TextDocument. Used for reporting.
  * New API: Reportable. This is an interface for anything that can provide a FileLocation (Nodes and Tokens currently)
  * New API: TextRegion. This represents a range of text in a TextDocument with offset/length, which is useful for tokens and for TextAvailableNodes in general 
  * changes to Node interface:
    * `getTextRegion` is a new abstract method. `getReportLocation` is not abstract anymore. TextAvailableNode refines the contract of `getTextRegion` to mean that the text region is the exact boundaries of the node in the text, which is eg not the case for the apex module, but is the case for Javacc-based modules.
 * Some usage of Path and File are replaced with String in eg Ruleset#applies (file name filters) because TextFile uses strings as path names.

* CPD
  * CPD uses FileLocation and TextDocument
  * The end goal is to replace SourceCode with TextDocument and TextFile
  * This refactoring is incomplete for now, some transitional APIs are in the class CpdCompat. Removing them requires #3919 
  * Antlr tokens are now easier to build because they can use a TextDocument
  
* pmd-java
  * AbstractJavaNode#getPreferredReportLocation is removed (Node#getReportLocation is the way now), ref #730 

* pmd-lang-test
  * Things to represent and test text locations are replaced with standard pmd-core APIs (FileLocation/TextRegion)

* pmd-test
  * Change to how the "code" element is trimmed, before it was just `String::trim()` which doesn't preserve indent on the first line. Now it uses basically a clone of `String::stripIndent` (method used by text blocks in JDK 16+) so that the code behaves like a text block (without escapes)

* pmd-cpp
  * The "skip block" feature is temporarily dysfunctional, it's fixed in the followup branch, I just ignored the tests

Future refactorings:
- violation decorators are needed to remove RuleViolationFactory
- finish the CPD refactoring (SourceCode -> TextDocument)
- integrate TextDocument more cleanly within our javacc support

## Related issues

- Part of #3784 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [ ] Passing all unit tests
- [ ] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)
- [x] Investigate required changes in the designer -> pmd/pmd-designer#51

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Running pmdtester failed, this message is mainly used to remind the maintainers of PMD.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java:189]: Hm... if we normalize e.g. `\r\n` into `\n`, then we change the text... Addressing content by linenumber + column should still be ok, but addressing by offset will be different. E.g. if we report the location of a violation, it might be rendered at a different position by tools because we analyzed not the original text, but a normalized text...
Won't this normalization create more problems than it solves?

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrToken.java
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrToken.java:24]: This looked much more complete than just `\r\n|\n`, which is used in TextFileContent...

**Review by oowekyala — COMMENTED** [review]
todos

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java:189]: Offsets are not output anywhere, they're valid on a given text document in memory, but are not absolute (eg with document views, the same place may have different offsets depending on whether you ask the view or its parent). We still only report violations with line/column. So I don't think that's a problem.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextDocument.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextDocument.java:38]: The translation of escape codes that happens before lexing in some languages. Eg in java these are the unicode escapes like `\u00a0`, in C++ the line continuations (`\` + newline). In the followup branch, they are decoded in a single pass over the original text document. That pass produces a new text document with the translated source (escapes replaced with the characters they stand for), which is then passed to the lexer/parser. Doing this in a separate pass simplifies the CharStream implementation for Javacc, and also makes sure that the text that is available in rules doesn't contain escapes. Otherwise you could eg  have a `class` token whose image is `\u0064lass`, so the `getText` method would be less useful

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/RootTextDocument.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/RootTextDocument.java:77]: Yeah I added TextPos2d much later than I wrote this (probably years later in real time ^^). I can refactor this.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/SourceCodePositioner.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/SourceCodePositioner.java:212]: This code handles `\r\n` correctly, as both `\r\n` and `\n` end with the same char. I realise it doesn't handle `\r`, I only very recently added support for that in TextFileContent. The charsequence is not expected to be normalized here so that code needs to be updated

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml
Comment (oowekyala) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml:299]: The thing is doc comments are comments, so I think this is correct. We could have a rule "dangling doc comment" that reports a doc comment that isn't before a class or method, but I don't think it's the job of EmptyCatchBlock. It's confusing to have the rule tell you to add a comment when there's already one there

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrToken.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrToken.java:24]: TextFileContent uses `\r\n?|\n`, so it handles three line terms. This is consistent with many Java core APIs like [`String#lines()`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html#lines()) or [`BufferedReader::readLine`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/BufferedReader.html#readLine())

File: pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java
Comment (adangel) [reply pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java:46]: Ah ok, `TextFileContent#getLineTerminator()` returns the _detected_ line terminator of the file and only falling back to System.lineSeparator() if the file is broken...

File: pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java
Comment (adangel) [inline pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java:109]: That's the old separator used by mac before macosx (unix) :smile:

File: pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFilesTest.java
Comment (adangel) [inline pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFilesTest.java:59]: That's unnecessary. Should we test for the display name maybe?

**Review by adangel — COMMENTED** [review]
So far I noticed the following API changes, that we probably should list somewhere...

Removed APIs:
*   net.sourceforge.pmd.document.* -> replaced by net.sourceforge.pmd.lang.document.* (TextDocument, ...)
*   net.sourceforge.pmd.lang.ast.SourceCodePositioner -> not needed anymore, it is implicitly included via TextDocument / TextFileContents

Changed/New APIs:
*   pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java
    *   DateSource -> TextFile (Renderers)
    *   Removed: `void setUseShortNames(List<String> inputPaths)`

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/Chars.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/Chars.java:448]: Thanks for these fixes!

File: pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java:333]: Yes, cf the javadoc: "Trailing whitespace is removed on each line."

File: pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java
Comment (oowekyala) [reply pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java:65]: I think both would be identical. The thing this is testing is that the Chars instance `normalizedText` has the same size as the underlying string, therefore, repeated application of toString will return that string instead of calling substring on the underlying string.

Comment (oowekyala) [reply pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFileContentTest.java:65]: I've tried to clarify this in comments

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/CpdCompat.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/CpdCompat.java:16]: Since I want to remove it before 7.0.0, I think it's correct not to mark it `@DeprecatedUntil700`. My understanding is that this annotation marks an API that should not be removed before 7.0.0-beta (or however we call the release that has these compatibility apis)

File: pmd-core/src/main/java/net/sourceforge/pmd/internal/util/BaseCloseable.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/internal/util/BaseCloseable.java:10]: will do while fixing #2873

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/SourceCodePositioner.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/SourceCodePositioner.java:201]: I think in fact, the charsequence is expected to be normalized here... I changed the visibility to package private

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/TextFileContent.java:189]: > only report violations with line/column

I think, that's key. It's maybe something to keep in mind - PMD won't be able to support a renderer that wants to output offset coordinates. I've checked the SARIF format - there both coordinate systems are supported, so we are fine here.

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml
Comment (adangel) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml:299]: I think, we should open an issue for this and ask @rsoesemann and @SCWells72 and other apex devs for input. The current state is, that EmptyCatchBlock doesn't consider doc comments as comments. It was modified together with #3569 . Maybe it's a rule bug, but then it should be explicit IMHO and not fixed as a side effect...

The current test data for the rule: https://github.com/pmd/pmd/blob/master/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrToken.java
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/antlr4/AntlrToken.java:24]: :+1: Yes, I think it's good to be consistent.

File: pmd-core/src/main/java/net/sourceforge/pmd/lang/document/CpdCompat.java
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/lang/document/CpdCompat.java:16]: ok, I guess it's anyway internal and no use as an API :)

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java:140]: Reminder for me -> Still need to rebase and push those changes (https://github.com/apache/maven-pmd-plugin/compare/master...adangel:pmd-6.45.0-SNAPSHOT)

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml
Comment (adangel) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml:299]: Created an issue: #3953

**Review by adangel — APPROVED** [review]
I leave merging this up to you.
Thanks for the work!

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/SingleLineComment.java
Comment (oowekyala) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/SingleLineComment.java:19]: The comments are not children of any java node in the Java AST. In this PR Comment is not a subclass of AbstractJavaNode anymore, so this method has been removed with no replacement (was dead code). I think we might need to change that back in the future if we want the designer to display these comments, and also access them in XPath

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java
Comment (oowekyala) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java:140]: Should I wait until you have setup this compat branch, or rather just change the version to the snapshot in this PR?

File: pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFilesTest.java
Comment (oowekyala) [reply pmd-core/src/test/java/net/sourceforge/pmd/lang/document/TextFilesTest.java:59]: I want to test the equals method... but `tf == tf` so indeed this is useless. I think we need another "tf" instance created to be equal but a different instance

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java:140]: According to the last build (https://github.com/pmd/pmd/actions/runs/2250604471) the dogfood run was executed without failure - so despite my fear, this doesn't seem to be a problem.

**But**: We absolutely have to fix regression tester first, see also my comment on #3892

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/SingleLineComment.java
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/SingleLineComment.java:19]: Ok, that explains it: dead code :)

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java
Comment (adangel) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/Renderer.java:140]: I've just now pushed a merge to this PR - hopefully the regression tester runs now without a problem.... :crossed_fingers:

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml
Comment (oowekyala) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml:299]: I'll try to revert this but IIRC this wasn't trivial, which is why I changed this

I'll try to make it so we can revert it later if needed

**Comment by adangel** [comment]
Ok, the regression tester ran now... but with some more work to do: It seems, the reported file names are now different.... all violations have been removed and added.

It's easier to look at a smaller project, like https://chunk.io/pmd/23054915090a48adb1f094dcfc0e981a/diff1/Schedul-o-matic-9000/index.html

The xml report (baseline): https://chunk.io/pmd/23054915090a48adb1f094dcfc0e981a/diff1/Schedul-o-matic-9000/base_pmd_report.xml

```xml
<file name="/home/runner/work/pmd/target/repositories/Schedul-o-matic-9000/force-app/main/default/classes/Dao.cls">
<violation beginline="8" endline="8" begincolumn="14" endcolumn="16" rule="ApexSharingViolations" ruleset="Security" externalInfoUrl="https://pmd.github.io/pmd/pmd_rules_apex_security.html#apexsharingviolations" priority="3">
Apex classes should declare a sharing model if DML or SOQL/SOSL is used
</violation>
...
```

The xml report (patch): https://chunk.io/pmd/23054915090a48adb1f094dcfc0e981a/diff1/Schedul-o-matic-9000/patch_pmd_report.xml

```xml
<file name="target/repositories/Schedul-o-matic-9000/force-app/main/default/classes/Dao.cls">
<violation beginline="8" endline="8" begincolumn="14" endcolumn="17" rule="ApexSharingViolations" ruleset="Security" externalInfoUrl="https://pmd.github.io/pmd/pmd_rules_apex_security.html#apexsharingviolations" priority="3">
Apex classes should declare a sharing model if DML or SOQL/SOSL is used
</violation>
...
```

The filenames are now relativized already... but only to the current working directory (probably).

**Comment by adangel** [comment]
I think, both behaviors make sense. But since we are on PMD 7, we can change it...

>  I think the good thing about this is that the report is maybe more portable.

:+1:

Hm... theoretically, we don't need to change the regression tester. It would work after this PR is merged, then the baseline also has the same paths in the report...

File: pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml
Comment (oowekyala) [reply pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/errorprone/xml/EmptyCatchBlock.xml:299]: Yeah so I tried but this is a lot of work, unless we're fine with saying that doc comments are not comments in Apex, which I think is bad. Let's wait for input on #3953

**Pull request merged**
